addons/stock/models/stock_move.py

```diff
@@ -495,13 +495,20 @@ def _set_lot_ids(self):
                     move_lines_commands.append((0, 0, move_line_vals))
             move.write({'move_line_ids': move_lines_commands})
 
-    @api.depends('picking_type_id', 'date')
+    @api.depends('picking_type_id', 'date', 'priority')
     def _compute_reservation_date(self):
         for move in self:
-            if move.state not in ['draft', 'confirmed', 'waiting', 'partially_available']:
+            if move.state not in ['draft', 'confirmed', 'waiting', 'partially_available'] or not move.picking_type_id:
                 continue
-            if move.picking_type_id and move.picking_type_id.reservation_method == 'by_date':
-                move.reservation_date = fields.Date.to_date(move.date) - timedelta(days=move.picking_type_id.reservation_days_before)
+            if move.picking_type_id.reservation_method == 'by_date':
+                days = move.picking_type_id.reservation_days_before
+                if move.priority == '1':
+                    days = move.picking_type_id.reservation_days_before_priority
+                move.reservation_date = fields.Date.to_date(move.date) - timedelta(days=days)
+            elif move.picking_type_id.reservation_method == 'at_confirm':
+                move.reservation_date = fields.Date.today()
```

Comment:
Set the `reservation_date` at the creation instead of at the actual validation ?
If I create a receipt A the 1st April but confirm it the 8 April, when it will be confirmed, its move still have the `reservation_date` at the 1st April. 
So, if between this time I created and validate a receipt B the 2 April, in the report receipt B will be after receipt A even if receipt A was confirmed after receipt B.
Is it wanted ?

Reply:
Hm, I forget why I originally decided this. My commit message does point out this is the case to avoid having `state` be part of the depends for performance reasons, but I'm not sure why I didn't add it as a write to `_action_confirm`. Does this seem like a good idea though?

Reply:
I'm not sure... Maybe it doesn't worth to add that in the `write` as the gain of precision seems trivial (it seems to have an impact only for the order in the forecast report, but I don't think it will create issue with the reservation).
I would prefer to have an additional opinion on that.

Reply:
Updated it so now `_action_confirm` sets the `reservation_date` for the 'at_confirm' moves and it will be `False` before that.